caput/weighted_median.pyx

```diff
@@ -9,99 +9,366 @@ from MedianTree cimport Tree, Data
+
+    # NOTE: this is super annoying. In theory this is what Cython fused types are meant
+    # to be for, but in practice it seems to be impossible to combine arbitrarily sized
+    # arrays (which require you to treat the arrays as Python objects), and fused
+    # types. Explicitly dispatching to the correct routines seems to be the lesser of
+    # all evils here
+    if A.dtype == np.intc:
+        if W.dtype == np.intc:
+            _quickselect[int, int](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.int:
+            _quickselect[int, long](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.single:
+            _quickselect[int, float](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.double:
+            _quickselect[int, double](Ar, Wr, q, methodc, res, At, Wt)
+        else:
+            raise TypeError("Type of weight array is not supported.")
+    elif A.dtype == np.int:
+        if W.dtype == np.intc:
+            _quickselect[long, int](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.int:
+            _quickselect[long, long](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.single:
+            _quickselect[long, float](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.double:
+            _quickselect[long, double](Ar, Wr, q, methodc, res, At, Wt)
+        else:
+            raise TypeError("Type of weight array is not supported.")
+    elif A.dtype == np.single:
+        if W.dtype == np.intc:
+            _quickselect[float, int](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.int:
+            _quickselect[float, long](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.single:
+            _quickselect[float, float](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.double:
+            _quickselect[float, double](Ar, Wr, q, methodc, res, At, Wt)
+        else:
+            raise TypeError("Type of weight array is not supported.")
+    elif A.dtype == np.double:
+        if W.dtype == np.intc:
+            _quickselect[double, int](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.int:
+            _quickselect[double, long](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.single:
+            _quickselect[double, float](Ar, Wr, q, methodc, res, At, Wt)
+        elif W.dtype == np.double:
+            _quickselect[double, double](Ar, Wr, q, methodc, res, At, Wt)
+        else:
+            raise TypeError("Type of weight array is not supported.")
+    else:
+        raise TypeError("Type of data array is not supported.")
 
-    if data.ndim is 1:
-        return _weighted_median_1D(data, weights, c_method)
-    if data.ndim is 2:
-        return _weighted_median_2D(data, weights, c_method)
-    raise NotImplementedError('weighted_median is only implemented for 1 and 2 dimensions, not {}'
-                              .format(data.ndim))
+    if A.ndim == 1:
+        return res[0]
```

Comment:
I don't understand what this case is about.

Reply:
It's so that the routine returns a scalar if the input array was only 1D.

Reply:
Oh yeah, it does make sense :+1: